src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/SemanticTokensCache.cs

```diff
@@ -0,0 +1,210 @@
+
+            // Keep track of where the cached range begins
+            Position? cachedRangeStart = null;
+
+            // Keep track of where the cached range ends
+            var cachedRangeNumLines = 0;
+
+            // Keeps track of the current empty line count (needed for calculating line offsets)
+            var emptyLineCount = 0;
+
+            // Retrieve the cached tokens associated with the passed-in range
+            for (var line = requestedRange.Start.Line; line < requestedRange.End.Line; line++)
+            {
+                if (!lineToTokensDict.TryGetValue(line, out var lineTokens))
+                {
+                    if (tokens.Count != 0)
+                    {
+                        // We already have some sort of cached results, but we've now reached
+                        // a gap in the cache which we have no tokens for.
+                        // Since we only want to return a continuous range, we'll exit early.
+                        break;
+                    }
+
+                    continue;
+                }
+
+                if (cachedRangeStart is null)
+                {
+                    cachedRangeStart = new Position { Line = line, Character = 0 };
+                }
+
+                cachedRangeNumLines++;
+
+                // If a line's List is empty, it means the line has no tokens
+                if (lineTokens.Count == 0)
+                {
+                    emptyLineCount++;
+                    continue;
+                }
+
+                // We've found a non-empty cached line. Go through and process it.
+                for (var lineTokenIndex = 0; lineTokenIndex < lineTokens.Count; lineTokenIndex++)
+                {
+                    // If we're at the start of the cached range, the line offset should be relative
+                    // to the start of the document.
+                    if (tokens.Count == 0)
+                    {
+                        tokens.Add(line);
+                    }
+                    // If we're not at start at the cached range but we're at the start of a line,
+                    // the line offset should be relative to the end of the last cached line.
+                    else if (lineTokenIndex == 0)
+                    {
+                        tokens.Add(1 + emptyLineCount);
+                    }
+                    else
+                    {
+                        tokens.Add(lineTokens[lineTokenIndex]);
+                    }
+                }
+
+                emptyLineCount = 0;
+            }
+
+            if (tokens.Count == 0)
+            {
+                // We couldn't find any tokens associated with the passed-in range
+                cachedTokens = null;
+                return false;
+            }
+
+            Assumes.NotNull(cachedRangeStart);
+
+            var endPosition = new Position { Line = cachedRangeStart.Line + cachedRangeNumLines, Character = 0 };
```

Comment:
Shoulder character be the end-position of the tokens or am I totally off (super possible 😆)?

Reply:
Ah, so this mimics how the LSP client sends us ranges. Instead of something like `Range[(0, 0), (100, 1000)]` they ask us for something like `Range[(0, 0), (101, 0)]`. For comparison's sake later on, I think it's easier to just match how they do it.